yang/frr-nexthop.yang

```diff
@@ -97,113 +108,98 @@ module frr-nexthop {
   }
 
   /*
-   * Nexthop object
+   * Common nexthop attributes grouping.
    */
-
-  grouping frr-nexthop {
+  grouping frr-nexthop-attributes {
     leaf nh-type {
       type nexthop-type;
       mandatory true;
       description
         "The nexthop type.";
     }
-
-    leaf gateway {
-      type gateway-address;
-      description
-        "The nexthop gateway address.";
-    }
-
     leaf vrf {
-      type string {
-        length "1..36";
-      }
+      type frr-vrf:vrf-ref;
       description
         "The nexthop vrf name, if different from the route.";
     }
-
+    leaf gateway {
+      type frr-nexthop:gateway-address;
+      description
+        "The nexthop gateway address.";
+    }
     leaf interface {
       type frr-interface:interface-ref;
       description
         "The nexthop egress interface.";
     }
-
     leaf bh-type {
+      when "../nh-type = 'blackhole'";
       type blackhole-type;
       description
         "A blackhole sub-type, if the nexthop is a blackhole type.";
     }
-
-    leaf flags {
-      type uint32;
-      description
-        "The nexthop's raw flags value.";
-    }
-
-    leaf is-duplicate {
-      type empty;
-      description
-        "Duplicate nexthop; will be ignored.";
-    }
-
-    leaf is-recursive {
-      type empty;
-      description
-        "Nexthop must be resolved through another gateway.";
-    }
-
-    leaf is-onlink {
-      type empty;
+    leaf onlink {
+      when "../nh-type = 'ip4-ifindex' or
+            ../nh-type = 'ip6-ifindex'";
+      type boolean;
+      default false;
       description
         "Nexthop is directly connected.";
     }
-
-    leaf is-active {
-      type empty;
-      description
-        "Nexthop is active.";
-    }
-
     uses rt-types:mpls-label-stack {
       description
         "Nexthop's MPLS label stack.";
     }
+  }
 
-    leaf mtu {
-      type uint32;
+  /*
+   * Single nexthop grouping.
+   */
+  grouping frr-nexthop {
+    container frr-nexthops {
       description
-        "The nexthop's specific MTU.";
+        "FRR nexthop object.";
+      list nexthops {
+        key "nh-type gateway interface";
```

Comment:
are the `gateway` and `interface` attributes always present?

Reply:
no for backhole or reject nexthops nh-type should be enough?

Reply:
so ... I thought someone said that yang required the `key` attributes to always be present?

Reply:
yes nh-type is always present.

Reply:
Same point as @mjstapp  all leafs speficied in the key need to be specified. Here three leafs are specified but as listed above based on the `nh-type` key like `gateway` or `interface` may not always present. 

RFC-7950:
```The combined values of all the leafs specified in the key are used to uniquely identify a list entry. All key leafs MUST be given values when a list entry is created. ```

Reply:
@chiragshah6 I have attached PDF with example of how we can do that before creating xpath.